hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/AbstractMemStore.java

```diff
@@ -376,6 +376,15 @@ ImmutableSegment getSnapshot() {
     return snapshot;
   }
 
+  @Override
+  public void close() {
+    // active should never be null
+    active.close();
+    // for snapshot, either it is empty, where we do not reference any real segment which contains a
```

Comment:
I'm of the opinion that we should always close all closeables. I don't dispute your comment here, but I feel like we should close the snapshot either way. This prevents future bugs where assumptions might be made about the presence of a close() method on ImmutableSegment.

Reply:
I was thinking the same way when implementing this patch but it is not easy to get things right. It requires a big refactoring, on how we do ref counting here. So I prefer we fix the problem first to see there are still other leaks, and open another issue for refactoring. The latter one will take some time, since the code is not easy to fully understand...

Reply:
Filed HBASE-28435 for revisiting the referenc counting way. Do you have any other concerns? @bbeaudreault